File: F#/SpaceAge.fs
module SpaceAge

// TODO: define the Planet type

type Planet =
    | Mercury
    | Venus
    | Earth
    | Mars
    | Jupiter
    | Saturn
    | Uranus
    | Neptune



let age (planet: Planet) (seconds: int64) : float =
    let secondsToEarthYears (secs: int64) = (float secs) / 31557600.0

    match planet with
    | Mercury -> secondsToEarthYears seconds / 0.2408467
    | Venus -> secondsToEarthYears seconds / 0.61519726
    | Earth -> secondsToEarthYears seconds
    | Mars -> secondsToEarthYears seconds / 1.8808158
    | Jupiter -> secondsToEarthYears seconds / 11.862615
    | Saturn -> secondsToEarthYears seconds / 29.447498
    | Uranus -> secondsToEarthYears seconds / 84.046846
    | Neptune -> secondsToEarthYears seconds / 164.79132

File: elixir/space_age.ex
defmodule SpaceAge do
  @earth_year 31_557_600

  @planet_orbital_period %{
    mercury: 0.2408467,
    venus: 0.61519726,
    earth: 1,
    mars: 1.8808158,
    jupiter: 11.862615,
    saturn: 29.447498,
    uranus: 84.016846,
    neptune: 164.79132
  }

  @type planet ::
          :mercury
          | :venus
          | :earth
          | :mars
          | :jupiter
          | :saturn
          | :uranus
          | :neptune

  @doc """
  Return the number of years a person that has lived for 'seconds' seconds is
  aged on 'planet', or an error if 'planet' is not a planet.
  """
  @spec age_on(planet, pos_integer) :: {:ok, float} | {:error, String.t()}
  def age_on(planet, seconds) when is_map_key(@planet_orbital_period, planet) do
    {:ok, (seconds / (@earth_year * @planet_orbital_period[planet])) |> format_years()}
  end

  def age_on(_planet, _seconds), do: {:error, "not a planet"}

  def format_years({:ok, year}), do: {:ok, year |> Float.round(2)}
  def format_years(v), do: v
end
